src/plugins/piral-translate/package.json

```diff
@@ -63,10 +63,12 @@
     "test": "echo \"Error: run tests from root\" && exit 1"
   },
   "dependencies": {
-    "deepmerge": "^4.2.2"
+    "deepmerge": "^4.2.2",
+    "get-value": "^3.0.1"
```

Comment:
I am not sure if we need this if all it does / helps us is to resolve `a.b` to `"a"."b"`. Instead, what we could do is to normalize the provided translations from having nested objects to be flat:

```json
{
  "a": "foo",
  "b": {
    "c": "bar",
    "d": "qxz"
  }
}
```

would become:

```json
{
  "a": "foo",
  "b.c": "bar",
  "b.d": "qxz"
}
```

This way the lookup would be just that - no additional dependency and no parsing / special lookup necessary.

Reply:
With a flat structure, many preceding parts of the keys must be defined redundantly:

**Flat**

```json
{
    "pages.a.table.a.columns.a.name": "a",
    "pages.a.table.a.columns.a.name": "b",
    "pages.a.table.a.columns.a.name": "c",
    "pages.a.table.a.columns.a.name": "d",
    "pages.a.table.a.columns.a.name": "e"
}
```

**Nested**

```json
{
    "pages": {
        "a": {
            "table": {
		"a": {
		    "columns": {
			"a": {
                            "name": "a"
                        },
			"b": {
                            "name": "b"
                        },
                        "c": {
                            "name": "c"
                        },
                        "d": {
                            "name": "d"
                        },
                        "e": {
                            "name": "e"
                        }
		    }
                }
	    }
	}
    }
}
```

The problem reminds me strongly of the non-existent nesting of CSS compared to SCSS.

For me, it's ok if only flat translations are still possible.

When setting the translations within the Pilet, we can flatten our translations in order to concatenate the keys with a dot.

Reply:
Well, my suggestion would have been that this flattening is happening automatically (i.e., by the code changes of this PR). I know that the key length will be higher, however, this is the usual trade-off of memory vs computation. I think that the translations / strings are read more often than written, so having this already pre-computed will yield a desired performance boost.

Reply:
All right, sorry, then we've talked past each other.

I will adjust the pull request accordingly 😊

Reply:
For a quick comparison see: https://jsben.ch/N7vmy

![image](https://github.com/smapiot/piral/assets/1766191/56c75528-b6f2-4a86-864f-2458271c05f9)

It's about 10 times faster for a nesting depth of 6. (for a nesting level of 2 it's still 6 times faster, see https://jsben.ch/Vm6WR)

Since rendering is the crucial part I'd like to keep this as fast as possible; hence consuming more memory up-front.

What we could also do is a hybrid approach (although this is the most complicated one). We start with the nested one and use `getValue` only as a fallback if nothing is found. However, if `getValue` returns something we add it to the "working" translation set. This way, one still pays the nesting cost, however, only once - and without having the immediately increased memory.

Reply:
I have changed the code.

Now the nested translations are transformed into flat translations during the assignment.

Reply:
The build fails because the new dependency (flat) requires Node >= 18.

I'll take it into account and then commit again. 😊

Reply:
Here is a version without any additional dependency (and seems to work great):

```ts
function flatten(target: Record<string, string>, source: any, prop: string) {
  if (typeof source === 'string') {
    target[prop] = source;
  } else if (typeof source === 'object' && source) {
    Object.keys(source).forEach(key => {
      flatten(target, source[key], prop ? `${prop}.${key}` : key);
    });
  }
}

function flat(obj: Record<string, any>): Record<string, string> {
  const target: Record<string, string> = {};
  return flatten(target, obj, '');
}

// Example:
const obj1 = {};
const obj2 = { a: 'foo', b: 'bar' }
const obj3 = { a: 'foo', b: { c: 'bar' }}
const obj4 = {a: {b:{c:'foo'}, d: 'bar'}, e: 'qxz', f: { g: 'bar2' }} 

flat(obj1) // {}
flat(obj2) // { a: "foo", b: "bar" }
flat(obj3) // { a: "foo", "b.c": "bar" }
flat(obj4) // { "a.b.c": "foo", "a.d": "bar", e: "qxz", "f.g": "bar2" }
```

Is this what we need or am I missing something?

Reply:
Many thanks for your code, it looks very good, I would have implemented it myself, but I'm busy with day-to-day business at the moment.

I'll add it tomorrow at the latest 😊

Reply:
Thanks again for your code 😊

I have implemented your changes and removed the dependency accordingly.

Reply:
Awesome - you rock!